File: backend/app/schemas/ai.py
"""Request and response bodies of the AI endpoints (section 6)."""

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import Field

from app.schemas.base import CLEAN, ApiModel, OutModel

Priority = Literal["low", "medium", "high", "urgent"]


class TaskSuggestionRequest(ApiModel):
    brief: str | None = Field(
        default=None,
        max_length=1000,
        pattern=CLEAN,
        description="What the work is about, in a sentence or two. Optional.",
        examples=["Launch the mobile onboarding flow"],
    )
    count: int = Field(default=5, ge=1, le=10, description="How many tasks to suggest.")


class MetaOut(OutModel):
    model: str | None
    prompt_version: str = Field(examples=["task-generation-v1"])
    request_id: UUID


class QuotaOut(OutModel):
    daily_limit: int
    remaining_today: int
    resets_at: datetime


class AiStatusOut(OutModel):
    enabled: bool
    features: list[str]
    quota: QuotaOut


class SuggestionOut(OutModel):
    title: str
    description: str
    priority: Priority
    due_in_days: int | None = Field(description="Days from today, 0 to 90; the interface adds it.")


class SuggestionsOut(OutModel):
    suggestions: list[SuggestionOut]
    meta: MetaOut


class RankedTaskOut(OutModel):
    task_id: UUID
    rank: int
    suggested_priority: Priority
    reason: str


class PrioritizationOut(OutModel):
    items: list[RankedTaskOut]
    meta: MetaOut


class AiSummaryOut(OutModel):
    summary: str
    risks: list[str]
    next_steps: list[str]
    meta: MetaOut

File: frontend/src/adapters/mock/ai.ts
import type { AiService } from "@/services/ai";

// A deterministic stand-in for tests and the mock scenarios: the same answers every time.
const META = {
  model: "mock-model",
  promptVersion: "mock-v1",
  requestId: "00000000-0000-4000-8000-000000000000",
};

export function createMockAi(): AiService {
  return {
    status: () =>
      Promise.resolve({
        enabled: true,
        features: ["task_generation", "prioritization", "project_summary"],
        quota: { dailyLimit: 20, remainingToday: 20, resetsAt: "2026-01-02T00:00:00Z" },
      }),
    suggestTasks: () =>
      Promise.resolve({
        suggestions: [
          {
            title: "Define the scope and goals",
            description: "Write down what is in scope.",
            priority: "high",
            dueInDays: 3,
          },
          {
            title: "Review progress with the team",
            description: "",
            priority: "low",
            dueInDays: null,
          },
        ],
        meta: META,
      }),
    prioritize: () => Promise.resolve({ items: [], meta: META }),
    summarize: () =>
      Promise.resolve({ summary: "The project is moving.", risks: [], nextSteps: [], meta: META }),
  };
}
